### PR TITLE
cwb3 3.4.33 (new formula)

### PR DESCRIPTION
**Pull request by schtepf**
IMS Open Corpus Workbench (CWB) is a set of tools for indexing and querying large text corpora with linguistic annotations, which is widely used in corpus linguistics, computational lexicography and related fields, e.g. via various CQPweb servers across the world. 

This new HomeBrew formula provides the preferred way of installing CWB on MacOS.

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/cwb3.rb
Comment (carlocab) [inline Formula/cwb3.rb:32]:
```suggestion
    args = %W[
      PLATFORM=homebrew-formula
      SITE=homebrew-formula
      FULL_MESSAGES=1
      PREFIX=#{prefix}
      HOMEBREW_ROOT=#{HOMEBREW_PREFIX}
    ]

    system "make", "all", *args
    ENV.deparallelize
    system "make", "install", *args
```

File: Formula/cwb3.rb
Comment (carlocab) [inline Formula/cwb3.rb:13]:
```suggestion
  homepage "https://cwb.sourceforge.io/"
  url "https://downloads.sourceforge.net/project/cwb/cwb/cwb-3.5-RC/cwb-3.4.33-src.tar.gz"
  sha256 "947d5f6d710fd8f818c2dc027e991b898308d581cb9b89c4ce6db054fbb20948"
  license "GPL-2.0-or-later"
  head "svn://svn.code.sf.net/p/cwb/code/cwb/trunk"
```
`head` needs a default branch. Probably something like
```ruby
head "svn://svn.code.sf.net/p/cwb/code/cwb/trunk", branch: "master"
```

File: Formula/cwb3.rb
Comment (carlocab) [inline Formula/cwb3.rb:51]: This should be declared after the dependencies.

File: Formula/cwb3.rb
Comment (carlocab) [inline Formula/cwb3.rb:97]:
```suggestion
      cppflags = Utils.safe_popen_read("#{bin}/cwb-config", "-I").strip.split
      ldflags = Utils.safe_popen_read("#{bin}/cwb-config", "-L").strip.split
      system ENV.cc, "-o", "test", *cppflags, "test.c", *ldflags
```
We should probably omit parenthesis to function calls in this test where possible for consistency with other formulae.

**Review by carlocab — COMMENTED** [review]
I think the Linux build fails because we don't package `libintl` on Linux (right, @Homebrew/linux?). Is it possible to avoid linking with `libintl` here (even if just on Linux)?

File: Formula/cwb3.rb
Comment (schtepf) [reply Formula/cwb3.rb:32]: Thanks for looking at this so quickly! I've just been investigating the issue myself.  A quick test run on my MacBook suggests that `-lintl` is no longer needed (it's from pkg-config on MacOS, but we don't actually seem to be using any of the code).  I will have to fix this in upstream and make a new source release after confirming that it's fine without `-lintl`.

**Question:** Should upstream use `pkg-config` to get the compiler and linker flags instead? I was worried that package names might not be the same in Linux, breaking the build.  Upstream also explicitly uses paths `-L $HOMEBREW_PREFIX/opt/...` so installed software doesn't break when upgrading the prerequisite libraries to newer versions.  When installed via formula `cwb3`, will HomeBrew ensure automatically that required versions of the prerequisites will remain available in Cellar?

File: Formula/cwb3.rb
Comment (schtepf) [reply Formula/cwb3.rb:13]: As far as I know, SVN doesn't have the concept of branches. I suppose it would be fine to specify a fake `master` branch if it is required here.

File: Formula/cwb3.rb
Comment (schtepf) [reply Formula/cwb3.rb:51]: I moved this into the test block so that it is only installed when needed for testing, but not (unnecessarily) for the compilation.  Fine to move, of course.

File: Formula/cwb3.rb
Comment (schtepf) [reply Formula/cwb3.rb:97]: Great, thanks for the help! I don't speak ruby …

File: Formula/cwb3.rb
Comment (schtepf) [reply Formula/cwb3.rb:32]: PS: I will need some guidance on how to submit a new pull request after we've resolved all these issues.

Comment (carlocab) [reply Formula/cwb3.rb]: Using `pkg-config` might be more portable. You _might_ still need `-lintl` to link with `glib` on macOS, though I think macOS `ld` is smart enough to work the recursive dependency out these days.

Are you concerned about `.pc` files referring to `Cellar` paths? That's not an issue on macOS because we rewrite install names to avoid referencing them. I believe it's also not an issue on Linux, but I'm not up-to-speed on the details.

There's no need to submit a new pull request -- you can just modify this one with `git commit --amend` and `git push --force`.

Comment (schtepf) [reply Formula/cwb3.rb]: Oh right, my bad. It _does_ link against `libintl`, but doesn't require the explicit flag.

There are two concerns about `pkg-config`:

- I don't have Linux Homebrew for testing, so I'm not sure whether `pkg-config libpcre` works there as well, or whether the `.pc` happens to have a different name on some distros. These package names seem to be entirely unsystematic (e.g. `libpcre` vs. `glib-2.0` vs. `ncurses`).
- The paths I get from `pkg-config` on MacOS point to the Cellar rather than opt/, e.g.

```
$ pkg-config --libs glib-2.0
-L/usr/local/Cellar/glib/2.70.2/lib -L/usr/local/opt/gettext/lib -lglib-2.0 -lintl
```

So whenever I upgrade to a new glib version, the Cellar directory is deleted and the installed programs fail to run.  Note that all paths are set in the upstream Makefile, so we'd have to dynamically patch the makefile if we have to rewrite the compiler/linker flags in the formula.

Comment (carlocab) [reply Formula/cwb3.rb]: `libpcre.pc` is what you get from `./configure && make install`, so `pkg-config libpcre` should work on Linux too. The package names for `pkg-config` should be the same on macOS and Linux (at least for Homebrew).

Does `cwb3` remember the flags from `pkg-config` invocations? As long as it doesn't do so explicitly then the Cellar references shouldn't matter (e.g. try `otool -D $(brew --cellar glib)/*/lib/libglib-2.0.dylib`). Remembering these paths explicitly is atypical -- otherwise we'd need to recompile dependents much more often every time a formula has its `version` or `revision` bumped.

Comment (schtepf) [reply Formula/cwb3.rb]: CWB simply passes the `-L` flags from `pkg-config` to `gcc` when linking.  This used to cause the described breakage on every `brew upgrade` and hence was changed to manually set paths to opt/.  Perhaps this has changed in the meantime.  Is there any way to install an older/different version of a formula to test whether the build process works with pkg-config now?

Comment (danielnachun) [reply Formula/cwb3.rb]: The functionality of `libintl` on Linux is provided internally by `glibc`, which is automatically linked to all compiled binaries, and should never be explicitly linked with `-lintl`.  On macOS the situation is more complex - I don't believe `libSystem` provides that automatically, so the flag is sometimes needed.

If you want to be able to test the build in Linux locally, you can use the Homebrew docker image we use in CI: ` docker run -it -m 16GB homebrew/ubuntu16.04:master`.  That already has Homebrew set up so the only thing you have to do is check out your branch on there and you can test in the same build environment we use in CI.

Comment (carlocab) [reply Formula/cwb3.rb]: We run `brew linkage` in CI to check that formulae aren't linking to libraries that no longer exist (this is part of the testing done for dependents when a formula is updated), so this shouldn't be too much of a concern.

> Is there any way to install an older/different version of a formula to test whether the build process works with pkg-config now?

With a bit of `git` hackery, sure. But it's not really straightforward, and probably not worth the time.

File: Formula/cwb3.rb
Comment (SMillerDev) [inline Formula/cwb3.rb:35]:
```suggestion
    # make sure default registry exists
    (HOMEBREW_PREFIX/"share/cwb/registry").mkpath
```

File: Formula/cwb3.rb
Comment (SMillerDev) [inline Formula/cwb3.rb:41]:
```suggestion
      CWB registry directory: #{default_registry}
```

File: Formula/cwb3.rb
Comment (schtepf) [reply Formula/cwb3.rb]: `pkg-config` indeed seems to work correctly now. Upstream has been fixed to use pkg-config where possible.

File: Formula/cwb3.rb
Comment (SMillerDev) [inline Formula/cwb3.rb:27]: Do these impact the build or are they just recorded somewhere? Because I would say platform should define macOS or Linux in most software.

Comment (schtepf) [reply Formula/cwb3.rb:27]: These make settings are configuration options for the upstream sourcecode, which doesn't use autoconf.
PLATFORM determines the entire build environment, not just darwin vs. linux.

**Comment by schtepf** [comment]
Is any more input needed from me or should I just wait for final review of the pull request?

File: Formula/cwb3.rb
Comment (carlocab) [inline Formula/cwb3.rb:44]:
```suggestion
  def default_registry
    HOMEBREW_PREFIX/"share/cwb/registry"
  end

  def post_install
    # make sure default registry exists
    default_registry.mkpath
  end

  def caveats
```
Let's not hardcode this in two places.

**Review by carlocab — COMMENTED** [review]
Just one more suggestion. Other than this, LGTM

File: Formula/cwb3.rb
Comment (schtepf) [reply Formula/cwb3.rb:44]: Excellent, thanks.

**Review by carlocab — APPROVED** [review]
Thanks!

**Pull request closed without merging**
